### PR TITLE
move fsdp handling to accelerate

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Moves PyTorch FSDP handling to Accelerate
2. Should be merged after #23151 
3. No user-facing change. Now, users can use `accelerate launch` for fsdp in Trainer, e.g.:
```
accelerate launch --num_processes=2 --use_fsdp --mixed_precision=bf16 --fsdp_auto_wrap_policy=TRANSFORMER_BASED_WRAP  --fsdp_transformer_layer_cls_to_wrap="BertLayer" --fsdp_sharding_strategy=1 --fsdp_state_dict_type=FULL_STATE_DICT ./examples/pytorch/text-classification/run_glue.py   --model_name_or_path bert-base-cased   --task_name $TASK_NAME   --do_train   --do_eval   --max_seq_length 128   --per_device_train_batch_size 16   --learning_rate 5e-5   --num_train_epochs 3   --output_dir /tmp/$TASK_NAME/ --overwrite_output_dir
```

Continue to use torchrun with trainer args as usual.
```
torchrun --nnodes 1 --nproc-per-node 2 ./examples/pytorch/text-classification/run_glue.py   --model_name_or_path bert-base-cased   --task_name $TASK_NAME   --do_train   --do_eval   --max_seq_length 128   --per_device_train_batch_size 16   --learning_rate 5e-5   --num_train_epochs 3   --output_dir /tmp/$TASK_NAME/ --overwrite_output_dir --fsdp "full_shard auto_wrap" --fsdp_transformer_layer_cls_to_wrap BertLayer --bf16
```

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2820,11 +2779,15 @@ def save_model(self, output_dir: Optional[str] = None, _internal_call: bool = Fa
             ShardedDDPOption.ZERO_DP_2 in self.args.sharded_ddp
             or ShardedDDPOption.ZERO_DP_3 in self.args.sharded_ddp
             or self.fsdp is not None
+            or getattr(self.accelerator.state, "fsdp_plugin", None) is not None
```
Comment (muellerzr) [inline src/transformers/trainer.py:2782]: Would these now be one in the same? (So we can use only the `fsdp_plugin` check, or `self.fsdp` for everything?)

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -432,7 +432,7 @@ class TrainingArguments:
                     - `"backward_pre"` : Prefetches the next set of parameters before the current set of parameter's
                       gradient
                         computation.
-                    - `"backward_pos"` : This prefetches the next set of parameters after the current set of
+                    - `"backward_post"` : This prefetches the next set of parameters after the current set of
```
Comment (muellerzr) [inline src/transformers/training_args.py:435]: Should this be a param we deprecate?

File: src/transformers/trainer.py (hunk 1 above)
Comment (muellerzr) [reply src/transformers/trainer.py:2782]: E.g. such as how we have the following below:

```python
if len(self.fsdp) > 0 and not self.fsdp_config["xla"]
```

**Review by muellerzr — APPROVED** [review]
Thanks! A few questions I had below

**Review by sgugger — COMMENTED** [review]
Should we deprecate the entire `fsdp_config` to tell users to use the Accelerate config instead? We wouldn't be able to actually remove it before a majot release, but we can remove the doc for instance.

File: src/transformers/trainer.py (hunk 1 above)
Comment (pacman100) [reply src/transformers/trainer.py:2782]: both are needed as FSDP XLA isn't supported in accelerate yet

**Comment by pacman100** [comment]
Hello Sylvain, we can't do that as FSDP XLA integration uses it and that isn't supported yet in accelerate

File: src/transformers/trainer.py (hunk 1 above)
Comment (muellerzr) [reply src/transformers/trainer.py:2782]: Makes sense, thanks :)

File: src/transformers/training_args.py (hunk 2 above)
Comment (pacman100) [reply src/transformers/training_args.py:445]: Hello, this isn't a param but the value for a config `fsdp_backward_prefetch`. there was a typo before. So, this is fine.

**Pull request merged**